### PR TITLE
Cooking with CAS

### PR DESCRIPTION
**Pull request by eggyknap**
Not sure how the commits not from me got in here, but I figure the pull request gets the point across well enough, without me trying to remove them from the history with whatever git-fu I can muster.

### CODE REVIEW

File: 2020/03/10/cooking-with-cas.html.md [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+One of our customers asked us to host a new suite of web-based applications for them and to protect them with a single sign-on (SSO) solution. Ok, easy enough;
+these applications were in fact designed with a particular SSO system in mind already, but our situation required a different done, and we eventually chose
+Apereo's open source [Central Authentication Server project](https://www.apereo.org/projects/cas), or CAS. I'd like to describe the conversion process we went
+through.
+
+## The ingredients
+
+Our customer's application suite included:
+
+* The principle Java application using [JAAS authentication](https://docs.oracle.com/javase/7/docs/technotes/guides/security/jaas/JAASRefGuide.html)
```
Comment (jonjensen) [inline 2020/03/10/cooking-with-cas.html.md]: s/principle/principal/

File: 2020/03/10/cooking-with-cas.html.md [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+## Configuring Wildfly Authentication
+
+The most important application in the suite depends on the JAAS-based security subsystem of the [Wildfly application server](https://www.wildfly.org/) it's
+deployed to. Originally it used a custom [LoginModule](https://docs.oracle.com/javase/7/docs/api/javax/security/auth/spi/LoginModule.html) that inspected
+request headers to find the ID of the authenticated user. Our first task was to configure our proxy server to remove this header from every request. We planned
+to disable the old authentication system entirely, of course, but this change ensured that even if we mistakenly left it enabled somewhere, malicious users
+couldn't exploit it for access.
+
+This application uses a declarative security policy: the deployment descriptor identifies a set of user roles, and another set of "security constraints". Each
+security constraint describes one or more URL patterns used in the application, and the set of user roles allowed to access URLs matching those patterns. Here's
```
Comment (jonjensen) [inline 2020/03/10/cooking-with-cas.html.md]: Here's: s/'s//

File: 2020/03/10/cooking-with-cas.html.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+configured a Spring
+[UserDetailsService](https://docs.spring.io/spring-security/site/docs/3.2.3.RELEASE/apidocs/org/springframework/security/core/userdetails/UserDetailsService.html)
+to execute the same database query we used above to find the user's roles. I'm not fully conversant in the large stack of beans Spring uses to manage the
+process, and it took some time to get this configuration sorted out.
+
+## Enter PHP
+
+Two of these applications use PHP, which meant yet another configuration. Apereo maintains a [PHP client](https://github.com/apereo/phpCAS), which includes
+several helpful examples. Once I refreshed my memory about how to use PHP itself, I tracked down the part of the application that authenticates users, and
+replaced the existing code with calls to php CAS:
```
Comment (jonjensen) [inline 2020/03/10/cooking-with-cas.html.md]: s/php CAS/phpCAS/ ?

File: 2020/03/10/cooking-with-cas.html.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+```php
+    php-CAS::client(CAS_VERSION_2_0, $phpCASHost, $phpCASHost, $phpCASContext);
```
Comment (jonjensen) [inline 2020/03/10/cooking-with-cas.html.md]: Is the hyphen in `php-CAS` right? Seems weird when package `phpCAS` is used on the very next line.

File: 2020/03/10/cooking-with-cas.html.md (hunk 1 above)
Comment (eggyknap) [reply 2020/03/10/cooking-with-cas.html.md]: Gah, I remember specifically thinking of that when writing something or other recently; I think it was this article. Oh well :)

File: 2020/03/10/cooking-with-cas.html.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+assertion comes from the CAS extension and not somewhere else. Then the `mapping` portion ([documented here](https://docs.wildfly.org/14/Admin_Guide.html#mapping))
+looks up the given user in a database to find what roles it should be assigned.
+
+The last piece of the puzzle is a CAS deployment descriptor for our application, which activates cas-extension for that application. In its simplest form, this
+is an empty file in the right place, but ours ended up a little more complex. It identifies both the CAS profile we want to use (unnecessary in this case, as
+there's only one CAS profile on the system, but it helped keep things more clear in our minds), and instructs the extension to load some other libraries into
+our application.
+
+```xml
+<?xml version="1.0" encoding="UT-8"?>
```
[the remark this answers is not in the document]
Comment (eggyknap) [reply 2020/03/10/cooking-with-cas.html.md]: This, and the `php-CAS` you noticed, both seemed like weird mistakes for me to have made, especially considering the code snippets were simply copied from other sources. Turns out I made some sort of spell check faux pas in an earlier commit. With help from you and from `git blame`, I tracked down a few other unwanted edits from an earlier commit. Thanks!

File: 2020/03/10/cooking-with-cas.html.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+describes the bulk of the configuration, which seemed less straightforward than for cas-extension, but follows essentially the same mechanism. Here we
+configured a Spring
+[UserDetailsService](https://docs.spring.io/spring-security/site/docs/3.2.3.RELEASE/apidocs/org/springframework/security/core/userdetails/UserDetailsService.html)
+to execute the same database query we used above to find the user's roles. I'm not fully conversant in the large stack of beans Spring uses to manage the
+process, and it took some time to get this configuration sorted out.
+
+## Enter PHP
+
+Two of these applications use PHP, which meant yet another configuration. Apereo maintains a [PHP client](https://github.com/apereo/phpCAS), which includes
+several helpful examples. Once I refreshed my memory about how to use PHP itself, I tracked down the part of the application that authenticates users, and
```
[the remark this answers is not in the document]
Comment (eggyknap) [reply 2020/03/10/cooking-with-cas.html.md]: Yeah, I had that thought myself, a little after I submitted the pull request. "Here, dear reader and potential customer: Let me tell you that I suck at PHP, and see if that contributes to our marketing efforts..."

File: 2020/03/10/cooking-with-cas.html.md (hunk 3 above)
Comment (eggyknap) [reply 2020/03/10/cooking-with-cas.html.md]: I blame that spell-check-gone-bad, again

**Pull request merged**